Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -258,7 +277,113 @@ TEST_F(SerializerTest, WritesNestedObjects) {
   ExpectRoundTrip(model, proto, FieldValue::Type::Object);
 }
 
+TEST_F(SerializerTest, BadNullValue) {
+  std::vector<uint8_t> bytes{
+      0x58,  // encoded null tag
+      0x01,  // invalid null value. (0 is only valid null value)
+  };
+  ExpectFailedStatusDuringDecode(
+      Status(FirestoreErrorCode::InvalidArgument, "ignored"), bytes);
+}
+
+TEST_F(SerializerTest, BadBoolValue) {
+  std::vector<uint8_t> bytes{
+      0x08,  // encoded bool tag
+      0x02,  // invalid value for a bool. (Valid values are 0,1)
+  };
+  ExpectFailedStatusDuringDecode(
+      Status(FirestoreErrorCode::InvalidArgument, "ignored"), bytes);
+}
+
+TEST_F(SerializerTest, BadIntegerValue) {
+  // clang-format off
+  std::vector<uint8_t> bytes{
+      0x10,  // encoded int tag
+      // These bytes represent a number too large to represent in a 64 bit
+      // value.  This should cause an overflow.
+      0xff, 0xff, 0xff, 0xff,
+      0xff, 0xff, 0xff, 0xff,
+      0xff, 0xff, 0x7f,
+  };
+  // clang-format on
+  ExpectFailedStatusDuringDecode(
+      Status(FirestoreErrorCode::InvalidArgument, "ignored"), bytes);
+}
+
+TEST_F(SerializerTest, BadStringValue) {
+  // clang-format off
+  std::vector<uint8_t> bytes{
+      0x8a, 0x01,  // encoded string tag
+      0x05,        // length 5
+      'a',
+  };
+  // clang-format on
+  ExpectFailedStatusDuringDecode(
+      Status(FirestoreErrorCode::InvalidArgument, "ignored"), bytes);
+}
+
+TEST_F(SerializerTest, BadTag) {
+  // The google::firestore::v1beta1::Value value_type oneof currently has tags
+  // up to 18. For this test, we'll pick a tag that's unlikely to be added in
+  // the near term but still fits within a uint8_t even when encoded.
+  // Specifically 31.
+  // clang-format off
+  std::vector<uint8_t> bytes{
+      0xF8, 0x01,  // represents field number 31 encoded as a varint
+                   // There's no payload here, which is also invalid, but we
+                   // won't get that far.
+  };
+  // clang-format on
+#if defined(NDEBUG)
```

Comment:
I'm pretty uncomfortable with this variation in here. Xcode does not define `NDEBUG` for release builds by default, though it does define `DEBUG` for debug builds. We should probably add some preprocessor code in a central header to normalize this.

Given that this is a non-production serializer I'd suggest making the unhandled case a `FIREBASE_ASSERT`. We can relax this later if needed.